keps/0006-lifecycle.md

```diff
@@ -0,0 +1,341 @@
+2. TestSuite control which includes; applying manifests, CRDs, and running commands.  This enables a "before" testsuite model, however it is limited to the full collection of TestSuites.
+
+#### Flat TestSuites
+
+Today, KUTTL treats all configured testsuites (listed with `testDirs`) as 1 suite of tests.  While it makes senses for all tests to run under the given control plane, it is limiting for a configured testsuite to not be able to express a set of commands or manifests which are specific to that set of tests.  It currently requires the test harness configuration change based on intimate knowledge of the tests.  It would be great to separate out these concerns, which will result in more portable tests.
+
+
+## Proposal
+
+### Proposal: Component Definition
+
+![TestSuite Architecture](images/kuttl-testsuite.jpg)
+
+* **TestSuite Collection** - This is a collection of TestSuites, which could be 1 or more.  This traditionaly has been called a TestSuite.  The TestSuite Collection is configured with the test harness configuration file, `kuttl.yaml` by default. 
+* **TestSuite** - Is a folder which contains a collection of Tests defined by sub-folders.
+* **Test** - Is a folder which contains a number of files which define the steps necessary to assert the correct of a test.  If a test folder has sub-folders they are not analyzed by KUTTL unless explicitly referenced through configuration or a teststep.
+* **TestStep** - Is the smallest component of a test, which are a collection of indexed files which govern their order of evaluation.  There are 4 basic types of TestStep files:
+   1. an apply manifest file (used for test setup),
+   2. an assert file,
+   3. an error file (used to assert the absence of a resource)
+   4. a TestStep file which can run commands, update timeouts and delete resources. 
+
+### Proposal: Control Files
+
+`kuttl.yaml` is the file that defines the KUTTL test collection AND the KUTTL test harness configuration.  For a period of time, it will be redundant with the existing `kuttl-test.yaml`.  As is the fact with `kuttl-test.yaml` today, it will NOT be required.  When this file is missing, the standard KUTTL defaults will be in affect or the command-line flag overrides.
+
+`kuttl-testsuite.yaml` is the file that defines the KUTTL testsuite.  It is detected by KUTTL by residing at the base of a testsuite folder.  It is not required.  When present it will configure the set of tests that make up the testsuite with new defaults and will run a set of capabilities before and after a testsuite.
+
+### Proposal: Component Types
+
+The `kuttl.yaml` defined above will contain a new type which will consist of test harness control, testing defaults and collection of test suites. Based on this new type, it is proposed to establish a `kuttl.dev/v1beta2` GVK which will define the deprecation cycle.
+
+```
+apiVersion: kuttl.dev/v1beta2
+kind: TestCollection
+
+beforeCollection:
```

Comment:
originally this was `before:` and for the context perhaps is sufficient.
I starting thinking that in the future we might as `beforeEach` and `before:` becomes less clear.
I could see value in an `afterEach` that does an assert after each testsuite (in the case of a test-collection) or after each test (in the case of a test-suite).

This KEP is getting large and I think holding of on adding more at this point makes sense... lets discuss if you disagree

Reply:
perhaps we can add it in the KEP... we don't have to implement the kep all at once.